### ISSUE TITLE
Monorepo: Remove esModuleInterop and allowSyntheticDefaultImports Options

### ISSUE THREAD

**Issue by holgerd77** [ethereumjs/ethereumjs-monorepo#1964]
As stated in the associated meta issue #1946 the [esModuleInterop](https://www.typescriptlang.org/tsconfig#esModuleInterop) and [allowSyntheticDefaultImports](https://www.typescriptlang.org/tsconfig#allowSyntheticDefaultImports) TypeScript compiler options should be set to `false` in our base `tsconfig.json` file, see reasoning linked in the meta issue.

This task should optimally be done before the Beta 1 release round, latest before Beta 2.

The options should be deactivated in succession respectively commit by commit with a successful CI run in between.

**Comment by acolytec3** [comment]
Doesn't this require ESM builds in order to complete?

**Comment by paulmillr** [comment]
It seems like in `ESM` you could still import typescript-compiled files even without `esModuleIntrop`. I've tested this in node.js v18 REPL. So the option seems useless.

**Comment by holgerd77** [comment]
> Doesn't this require ESM builds in order to complete?

Read up on what the options do (these only cover some special import cases being otherwise problematic), also on the reasoning mentioned in the original TODO from the meta issue.

**Comment by ScottyPoi** [comment]
I worked on this yesterday and got most of the tests passing.

These two options work together to compensate for differences in how various modules export themselves.  Modules that do not `export default` their *namespace* cannot be imported in the same way as those that do.  The two compiler options allow us to ignore this and import all *namespaces* using the same syntax.   With them both set to *true*, all of our imports can look like:
 ```
import { BlockHeader } from '@ethereumjs/block'
import PeerId from 'peer-id'
import bootnodes from './bootnodes.json'
 ```
With these two options set to *false* (or **default** by excluding them), certain imports would require different syntax:
 ```
import { BlockHeader } from '@ethereumjs/block'
import * as PeerId from 'peer-id'
import * as bootnodes from './bootnodes.json'
 ```
From [](https://www.typescriptlang.org/tsconfig#esModuleInterop) 

```
By default (with esModuleInterop false or not set) TypeScript treats CommonJS/AMD/UMD modules similar to ES6 modules. In doing this, there are two parts in particular which turned out to be flawed assumptions:

    a namespace import like import * as moment from "moment" acts the same as const moment = require("moment")

    a default import like import moment from "moment" acts the same as const moment = require("moment").default

This mis-match causes these two issues:

    the ES6 modules spec states that a namespace import (import * as x) can only be an object, by having TypeScript treating it the same as = require("x") then TypeScript allowed for the import to be treated as a function and be callable. That’s not valid according to the spec.

    while accurate to the ES6 modules spec, most libraries with CommonJS/AMD/UMD modules didn’t conform as strictly as TypeScript’s implementation.
```

In #1973 I went through the whole monorepo to change all affected imports.  This touched *232* files, mostly because it affected `import tape from 'tape'` in every test file.   All of the tests passed.  I introduced some kind of bug along the way that broke the client test, but I know that it passed locally at some point.  I know it has something to do with the syntax for importing a namespace and using it as a constructor, which also works differently without `esModuleInterop`.

I'm interested in your thoughts on this.  From what I gathered, the way our code looks with these options *off*  is **valid** typescript.  Functionally, though, it is easier to be able to ignore these things, and the code looks more uniform.  I do not yet understand the underlying benefit / drawback of compiling with or without options.

**Comment by holgerd77** [comment]
@ScottyPoi Thanks a lot for giving this a go.

I have to say I am getting a bit undecided on this again as well, since this is again one of those cases where it is extremely hard to judge on the benefits on some "quality" level, so in the sense "what is the extend of the benefits", to than compare and weight towards the downsides. So and on the downsides side you are very much right, our code is/looks definitely more stream-lined and clean with theses options on, also to note that on the TypeScript side having this option (but: not specifically stating a library-development-context, rather some whatever TypeScript project) is recommended to be activated.

I also have the arguments from the article here https://www.semver-ts.org/#module-interop, which sounds theoretically convincing, I nevertheless ask myself how practically relevant this really is.

For example "The options are viral: enabling them in a package requires all downstream consumers to enable them as well (even if this is not desirable for whatever reasons)"

-> Let's assume that a lot of people (projects) DO activate these options already? Wouldn't this mean that it might practically not possible for the very most of the projects to deactivate any more anyhow, since there are always a few of your dependencies down the line which have them activated? 🤔

Maybe we give this a couple of days more and do NOT yet include in the Beta 1 releases.

Should I e.g. ask on Twitter if someone has problems with these options?

Any other idea how to get to some more solid ground on this? Thing is also: this problem is hard to google (which might actually also be an indicator that it's not that big of a problem at all).

**Comment by paulmillr** [comment]
My arguments are as following:

1. We're making a breaking release now, and the PR is ready, the timing is great.
2. The `: true` adds default imports. However, this is not how ESM behaves by default. This is something custom. If we want to match ESM behavior, it should be disabled. ESM is the future, after all. It's a matter of time when most software switches to it.
3. Scotty changed the imports of many modules e.g. `tape` in the PR — the imports match ESM now. I don't think it's good to promote non-standard import practice. Better be more strict and specify exactly what you're importing.
4. All imports with `: true` are wrapped — which can cause small performance degradation
5. See [this discussion in sindre's tsconfig](https://github.com/sindresorhus/tsconfig/issues/8)

What are benefits of leaving the option enabled? It seems like backwards compatibility is the only one, however, the release already breaks a lot of stuff.

**Comment by holgerd77** [comment]
@paulmillr ok, I trust on your assessment, so let's continue here. Then let's optimally also include in Beta 1.

@ScottyPoi if possible it would be great if this can get finalized today so to be ready for approval and merge.

**Comment by ScottyPoi** [comment]
@paulmillr @holgerd77 

Thank you for your review.  I will clear up the last few issues here and finalize by end of day.

From my research I do agree with paul's position about embracing the future.
Also I believe downstream compatibility is only an issue if we have options set to `true` that the downstream package does not.  Having these options off is default, which should not be an issue one way or another downstream.

**Comment by holgerd77** [comment]
Closed by #1975
